### ISSUE TITLE
Fix S4004 FP: Should not raise when enforced by interface or base class

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#2606]
### Description
As classes that have to implement writable collection properties have no influence on changing the contract S4004 ( Collection properties should be read-only) should not be raised. Related to #2344.

### Expected behavior
``` C#
public class S4004Interface : IS4004
{
    public IDictionary<object, object> Items { get; set; }// Compliant, enforced by interface
}
public interface IS4004
{
    IDictionary<object, object> Items { get; set; } // non-compliant
}

public class S4004Abstract : S4004Base
{
    public override IDictionary<object, object> Items { get; set; } // Compliant, enforced by base
}
public abstract class S4004Base
{
    public abstract IDictionary<object, object> Items { get; set; } // non-compliant
}
```

### Related information
* SonarC# Version 7.16
